xbmc/addons/InputStream.cpp

```diff
@@ -0,0 +1,506 @@
+#include "utils/log.h"
+#include "cores/VideoPlayer/DVDDemuxers/DVDDemux.h"
+#include "utils/RegExp.h"
+
+namespace ADDON
+{
+
+std::unique_ptr<CInputStream> CInputStream::FromExtension(AddonProps props, const cp_extension_t* ext)
+{
+  std::string listitemprops = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@listitemprops");
+  std::string extensions = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@extension");
+  std::string name(ext->plugin->identifier);
+  return std::unique_ptr<CInputStream>(new CInputStream(std::move(props),
+                                                        std::move(name),
+                                                        std::move(listitemprops),
+                                                        std::move(extensions)));
+}
+
+CInputStream::CInputStream(AddonProps props, std::string name, std::string listitemprops, std::string extensions)
+: InputStreamDll(std::move(props))
+{
+  m_fileItemProps = StringUtils::Tokenize(listitemprops, "|");
+  for (auto &key : m_fileItemProps)
+  {
+    StringUtils::Trim(key);
+    key = name + "." + key;
+  }
+
+  m_extensionsList = StringUtils::Tokenize(extensions, "|");
+  for (auto &ext : m_extensionsList)
+  {
+    StringUtils::Trim(ext);
+  }
+}
+
+bool CInputStream::Supports(CFileItem &fileitem)
+{
+  std::string extension = URIUtils::GetExtension(fileitem.GetPath());
+  bool match = false;
+  for (auto &ext : m_extensionsList)
+  {
+    if (ext == extension)
+    {
+      match = true;
+      break;
+    }
+  }
+  if (!match)
+    return false;
+
+  if (!m_pStruct)
+    return true;
+
+  std::string pathList;
+  try
+  {
+    pathList = m_pStruct->GetPathList();
+  }
+  catch (std::exception &e)
+  {
+    return false;
+  }
+
+  m_pathList = StringUtils::Tokenize(pathList, "|");
+  for (auto &path : m_pathList)
+  {
+    StringUtils::Trim(path);
+  }
+
+  match = false;
+
+  for (auto &path : m_pathList)
+  {
+    if (path.empty())
+      continue;
+
+    CRegExp r(true, CRegExp::asciiOnly, path.c_str());
+    if (r.RegFind(path.c_str()) == 0 && r.GetFindLen() > 6)
+    {
+      match = true;
+      break;
+    }
+  }
+  return match;
+}
+
+bool CInputStream::Open(CFileItem &fileitem)
+{
+  INPUTSTREAM props;
+  props.m_nCountInfoValues = 0;
+  std::vector<std::string> values;
+  for (auto &key : m_fileItemProps)
+  {
+    if (fileitem.GetProperty(key).isNull())
+      continue;
+    props.m_ListItemProperties[props.m_nCountInfoValues].m_strKey = key.c_str();
+    values.push_back(fileitem.GetProperty(key).asString());
+    props.m_ListItemProperties[props.m_nCountInfoValues].m_strValue = values.back().c_str();
+    props.m_nCountInfoValues++;
+  }
+  props.m_strURL = fileitem.GetPath().c_str();
+
+  bool ret = false;
+  try
+  {
+    ret = m_pStruct->Open(props);
+    if (ret)
+      m_caps = m_pStruct->GetCapabilities();
+  }
+  catch (std::exception &e)
+  {
+    CLog::Log(LOGERROR, "CInputStream::Open - could not open stream");
```

Comment:
I agree that this can be improved but this is work for another day. If you or @mapfau wants to take it, that'll be great.